### ISSUE TITLE
Porting to android

### ISSUE THREAD

**Issue by fayaz-modz** [alacritty/alacritty#7121]
Hi there everyone,
I'm want to port to the terminal to android. As this, to my knowledge, is written in rust. It shouldn't be a so much big issue to port to android, as rust provides arm64 v8 as a target.

I'm a flutter developer although I have some knowledge of android. I want some basic starting point to be able to port this to android. Can anyone give me a roadmap or help me bringing this to android?

Thanks in advance.

**Comment by kchibisov** [comment]
All the libraries we're using including internal structure should work on Android just fine, however you'd need to figure out how to make `pty` infrastructure work.

Upstream library winit is still in process of making soft keyboard work https://github.com/rust-windowing/winit on Android, so while you could try to make it work, I'm not sure you'll be able to start it.

Right now though, you could start basically by targetting `Android` and see what works and what's not, all the windowing bits should just work.

For more real time contact me either on `#winit` matrix chat or `#alacritty` at libera chat.

The only hard part of porting could be a `PTY` interface, the rest should be fairly simple as in adding a bunch of cfgs.

**Comment by Notarin** [comment]
I'm interested in some specifications. How will keyboard functionality work? The default keyboard by itself is **not** enough for a usable terminal environment, are we going to construct keyboard ourselves? Write our own supplementary bar? Or maybe use Termux's or a Termux inspired one?
Also, there are other issues as well, wakelock?, multiple sessions? session handling? permissions handling?
Due to the nature of Android this seems like it would end up its own entire project to get decent UX.
On that note, would this ever even hold a candle to Termux?

Am I being overly skeptical or highly overestimating the difficulty of this task?

**Comment by chrisduerr** [comment]
Closing because I don't think there's any interest from Alacritty maintainers to try and implement this.

If someone else wants to give it a shot and it's simple enough to upstream, feel free. But I wouldn't hold my breath.

**Comment by InternetOfTofu** [comment]
Hi,
I got it working. I'm still in the process of cleaning up and polishing. A screenshot is attached. I'll post the patch in a fork as soon as I can, but I want to reach out first.

![image](https://github.com/user-attachments/assets/580f748c-85e1-41a9-918f-0433f6d2a615)

What works from current Alacritty codebase, what's not, and what I tried to fix it:
(with all the dependencies sorted out)
1. Keyboard handling has a lot of differences. `winit::Keyboard::KeyEvent` from Android does not have extension traits implemented, as Android seemingly don't have scan code. The `KeyEvent` also don't contain `.text` field at all, but it has `PhysicalKey` and `LogicalKey`. I did some hacking to parse these and get the basic input compiled and working. I haven't even start to think about how to add key combination & binding support yet.
2. Font. As Android don't have fontconfig, it also doesn't have fontconfig config files / fonts file folder readiliy available. This will cause Alacritty to crash on startup. I worked around it currently by putting such folders copied from Ubuntu into "sdcard" and hardcoded environment variables in the Android code to point fontconfig to the right place. So a patch to add Android support to `crossfont` is a prerequisite for an Android Alacritty port to work. 
3. The graphics is working correctly, which is awesome. pty also seem to work just fine. 
4. Window resizing and text selection with touching works too. A physical keyboard is required for input.

The current port roughly works like this:
- a conanfile.py is used to compile the native dependencies (e.g. `libxkbcommon`, `libfontconfig`, `libfreetype`, etc.) and can produce `*.pc` pkg-config files for the `-sys` deps to consume
- `winit`'s `android-native-activiy` feature is used, and some mostly clean addition are added to paths related to windowing. A `lib.rs` is added as Android activity's entry point 
- The proejct compiles into an Android native activity with `cargo ndk`, and when using `cargo-mobile`'s example project and setting this native activity as entrypoint will launch it on Android. 

I wonder what type of project structure does the maintainers want if a mostly clean Android platform support is possible? What I'm thinking is having all rust code with `target_os = "android"` upstreamed in `alacritty` and `crossfont` repositories, and keeping the Android project directory in a separate repository, with a separate Android activity in that for support of e.g. config file & font loading, and interacting with Android framework for things like permission. 

For the discussion on the potential use mentioned by @kchibisov  : I'm try targeting Android large screen devices such as tablets and foldables. Alacritty can be a great desktop-grade terminal on these devices and function at least as an ssh client and possibly a lot more (e.g. with things like https://www.androidauthority.com/android-linux-terminal-app-3489887/)
